src/main/java/com/google/sps/servlets/TripsServlet.java

```diff
@@ -87,47 +88,151 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-        JsonObject curr = locationIterator.next().getAsJsonObject();
-        String placeID = curr.getAsJsonPrimitive("id").getAsString();
-        String placeName = curr.getAsJsonPrimitive("name").getAsString();
-        int weight = curr.getAsJsonPrimitive("weight").getAsInt();
-        TripLocation location = TripLocation.create(placeID, placeName, weight, userEmail);
-        datastore.put(convertTripLocationToEntity(location, tripEntity));
-      }
+    JsonArray locationData = jsonObject.getAsJsonArray("locations");
+    Iterator<JsonElement> locationIterator = locationData.iterator();
 
-      response.setStatus(HttpServletResponse.SC_OK);
-    } else {
-      // send error code 401 if user is not authenticated and tries to access data
+    while(locationIterator.hasNext()) {
+      TripLocation location = convertJsonObjectToTripLocation(locationIterator.next().getAsJsonObject(), userEmail);
+      datastore.put(convertTripLocationToEntity(location, tripEntity));
+    }
+
+    response.setStatus(HttpServletResponse.SC_CREATED);
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+      return;
+    }
+    String userEmail = userService.getCurrentUser().getEmail();
+    JsonObject jsonObject = BodyParser.parseJsonObjectFromRequest(request);
+    long timestamp = jsonObject.getAsJsonPrimitive(Trip.ENTITY_PROPERTY_TIMESTAMP).getAsLong();
+    
+    Trip trip = convertJsonObjectToTrip(jsonObject, userEmail, timestamp);
+    Entity tripEntity = convertTripToEntity(trip);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Query tripQuery = new Query("trip")
+                        .setFilter(new FilterPredicate(Trip.ENTITY_PROPERTY_OWNER, FilterOperator.EQUAL, userEmail))
+                        .setFilter(new FilterPredicate(Trip.ENTITY_PROPERTY_TIMESTAMP, FilterOperator.EQUAL, timestamp));
+    
+    Iterable<Entity> tripEntityIterable = datastore.prepare(tripQuery).asIterable();
+    if (Iterables.size(tripEntityIterable) == 0) {
+      /* If we reach this breakpoint, that means no Trip matches the timestamp 
+          to be updated. Send back an error code 404. */
+      response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+      return;
+    }
+    Entity existingTrip = tripEntityIterable.iterator().next();
+    Query tripLocationsQuery = new Query("trip-location")
+                                .setAncestor(existingTrip.getKey());
+    Iterable<Entity> tripLocationEntityIterable = datastore.prepare(tripLocationsQuery).asIterable();
+
+    // delete current data relating to Trip
+    datastore.delete(existingTrip.getKey());
+    for (Entity tripLocationEntity : tripLocationEntityIterable) {
+      datastore.delete(tripLocationEntity.getKey());
+    }
```

Comment:
It is not common in real world to delete and re-add the entry for each update. Is it a workaround for the bug in the other comment? My intuition is to "upsert" the trip location, delete removed ones, and update the trip. 

Also, in either way, consider to put them in a transaction(https://cloud.google.com/datastore/docs/concepts/transactions#using_transactions). Any of the database operation may fail and it may cause inconsistent trip data.

Reply:
My original implementation strategy was to upsert the trip, which is why I opted to use a PUT request for everything. Actually, in all my DB "insertion" methods, I am using put which is an upsert function.

The problem I encountered was that there was no trivial way to set a key of an Entity while creating it, and to make an update happen. I tried to use Entity Builders, but the current version of Datastore API in the classpath doesn't support it.

I will try to refactor this in a later PR, since it would be a performance boost to take the set intersection of trip locations and delete the set difference, while updating the Trip object.

I will implement the transactions for every insertion/"update" database operation in this PR though.